cardano-crypto-class/src/Cardano/Crypto/PinnedSizedBytes.hs

```diff
@@ -0,0 +1,229 @@
+    ws = reverse
+        $ take size
+        $ (++ repeat 0)
+        $ reverse ws0
+
+-- This is not efficient, but we don't use this in non-tests
+psbFromByteString :: KnownNat n => BS.ByteString -> PinnedSizedBytes n
+psbFromByteString = psbFromBytes . BS.unpack
+
+psbFromByteStringCheck :: forall n. KnownNat n => BS.ByteString -> Maybe (PinnedSizedBytes n)
+psbFromByteStringCheck bs 
+    | BS.length bs == size = Just $ unsafeDupablePerformIO $
+        BS.useAsCStringLen bs $ \(Ptr addr#, _) -> do
+            marr@(MutableByteArray marr#) <- newPinnedByteArray size
+            primitive_ $ copyAddrToByteArray# addr# marr# 0# (case size of I# s -> s)
+            arr <- unsafeFreezeByteArray marr
+            return (PSB arr)
+    | otherwise            = Nothing
+  where
+    size :: Int
+    size = fromInteger (natVal (Proxy :: Proxy n))
+
+psbZero :: KnownNat n =>  PinnedSizedBytes n
+psbZero = psbFromBytes []
+
+instance KnownNat n => Storable (PinnedSizedBytes n) where
+    sizeOf _          = fromInteger (natVal (Proxy :: Proxy n))
+    alignment _       = alignment (undefined :: FunPtr (Int -> Int))
+
+    peek (Ptr addr#) = do
+        let size :: Int
+            size = fromInteger (natVal (Proxy :: Proxy n))
+        marr@(MutableByteArray marr#) <- newPinnedByteArray size
+        primitive_ $ copyAddrToByteArray# addr# marr# 0# (case size of I# s -> s)
+        arr <- unsafeFreezeByteArray marr
+        return (PSB arr)
+
+    poke p (PSB arr) = do
+        let size :: Int
+            size = fromInteger (natVal (Proxy :: Proxy n))
+        copyByteArrayToAddr (castPtr p) arr 0 size
+
+psbUseAsCPtr :: PinnedSizedBytes n -> (Ptr Word8 -> IO r) -> IO r
+psbUseAsCPtr (PSB ba) k = k (byteArrayContents ba)
+
+psbUseAsSizedPtr :: PinnedSizedBytes n -> (SizedPtr n -> IO r) -> IO r
+psbUseAsSizedPtr (PSB ba) k = k (SizedPtr $ castPtr $ byteArrayContents ba)
+
+psbCreate :: forall n. KnownNat n => (Ptr Word8 -> IO ()) -> IO (PinnedSizedBytes n)
+psbCreate k = do
+    let size :: Int
+        size = fromInteger (natVal (Proxy :: Proxy n))
+    mba <- newPinnedByteArray size
+    k (mutableByteArrayContents mba)
+    arr <- unsafeFreezeByteArray mba
+    return (PSB arr)
+
+psbCreateSized :: forall n. KnownNat n => (SizedPtr n -> IO ()) -> IO (PinnedSizedBytes n)
+psbCreateSized k = psbCreate (k . SizedPtr . castPtr)
+
+ptrPsbToSizedPtr :: Ptr (PinnedSizedBytes n) -> SizedPtr n
+ptrPsbToSizedPtr = SizedPtr . castPtr
+
+-------------------------------------------------------------------------------
+-- derivative from primitive
+-------------------------------------------------------------------------------
+
+-- | Create a 'ByteArray' from a list of a known length. If the length
+--   of the list does not match the given length, this throws an exception.
+pinnedByteArrayFromListN :: forall a. Prim.Prim a => Int -> [a] -> ByteArray
+pinnedByteArrayFromListN n ys = runST $ do
+    marr <- newPinnedByteArray (n * Prim.sizeOf (head ys))
```

Comment:
We should at least document that this will error for empty lists. `pinnedByteArrayFromListN 0 []` will fail :-)

Reply:
I'd actually prefer a more robust solution. I'll make it `die` explicitly for now, but it might be prudent to lift this constraint to the type level, using non-empty list types.

Reply:
Or we could just return a zero-length byte array in that case? Thoughts?

Reply:
This is private function used in `psbFromBytes`. Don't overthink it.